File: spring-web-java/src/main/java/spring/web/java/infrastructure/security/JwtTokenProvider.java
package spring.web.java.infrastructure.security;

import java.util.Date;

import javax.crypto.SecretKey;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.io.Decoders;
import io.jsonwebtoken.security.Keys;
import spring.web.java.domain.model.enums.MemberRole;

public class JwtTokenProvider {

	private final SecretKey accessTokenKey;
	private final long accessTokenExpirationTime;
	private final SecretKey refreshTokenKey;
	private final long refreshTokenExpirationTime;

	public JwtTokenProvider(
		String accessTokenKey,
		long accessTokenExpirationTime,
		String refreshTokenKey,
		long refreshTokenExpirationTime
	) {
		this.accessTokenKey = Keys.hmacShaKeyFor(Decoders.BASE64.decode(accessTokenKey));
		this.accessTokenExpirationTime = accessTokenExpirationTime * 60 * 1000;
		this.refreshTokenKey = Keys.hmacShaKeyFor(Decoders.BASE64.decode(refreshTokenKey));
		this.refreshTokenExpirationTime = refreshTokenExpirationTime * 60 * 1000;
	}

	public String createAccessToken(Long memberId, MemberRole role) {
		return Jwts.builder()
			.claim("memberId", memberId)
			.claim("role", role)
			.issuedAt(new Date())
			.expiration(new Date(new Date().getTime() + accessTokenExpirationTime))
			.signWith(accessTokenKey)
			.compact();
	}

	public String createRefreshToken() {
		return Jwts.builder()
			.issuedAt(new Date())
			.expiration(new Date(new Date().getTime() + refreshTokenExpirationTime))
			.signWith(refreshTokenKey)
			.compact();
	}

	public Claims parseAccessToken(String token) {
		return Jwts.parser()
			.verifyWith(accessTokenKey)
			.build()
			.parseSignedClaims(token)
			.getPayload();
	}

	public Claims parseRefreshToken(String token) {
		return Jwts.parser()
			.verifyWith(refreshTokenKey)
			.build()
			.parseSignedClaims(token)
			.getPayload();
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/global/config/JwtTokenProvider.kt
package spring.web.kotlin.global.config

import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.io.Decoders
import io.jsonwebtoken.security.Keys
import spring.web.kotlin.domain.member.Member
import java.util.*
import javax.crypto.SecretKey

class JwtTokenProvider(
    private val accessTokenKey: SecretKey,
    private val accessTokenExpirationTime: Long,
    private val refreshTokenKey: SecretKey,
    private val refreshTokenExpirationTime: Long
) {
    constructor(
        accessTokenKey: String,
        accessTokenExpirationTime: Long,
        refreshTokenKey: String,
        refreshTokenExpirationTime: Long
    ) : this(
        accessTokenKey = Keys.hmacShaKeyFor(Decoders.BASE64.decode(accessTokenKey)),
        accessTokenExpirationTime = accessTokenExpirationTime * 60 * 1000,
        refreshTokenKey = Keys.hmacShaKeyFor(Decoders.BASE64.decode(refreshTokenKey)),
        refreshTokenExpirationTime = refreshTokenExpirationTime * 60 * 1000
    )

    fun createAccessToken(memberId: Long, role: Member.Role): String =
        Jwts.builder()
            .claim("memberId", memberId)
            .claim("role", role)
            .issuedAt(Date())
            .expiration(Date(Date().time + accessTokenExpirationTime))
            .signWith(accessTokenKey)
            .compact()

    fun createRefreshToken(): String =
        Jwts.builder()
            .issuedAt(Date())
            .expiration(Date(Date().time + refreshTokenExpirationTime))
            .signWith(refreshTokenKey)
            .compact()

    fun parseAccessToken(token: String): Claims =
        Jwts.parser()
            .verifyWith(accessTokenKey)
            .build()
            .parseSignedClaims(token)
            .payload

    fun parseRefreshToken(token: String): Claims =
        Jwts.parser()
            .verifyWith(refreshTokenKey)
            .build()
            .parseSignedClaims(token)
            .payload
}